File: 1021. Remove Outermost Parentheses/_1021.py
class Solution:
    def removeOuterParentheses(self, S: str) -> str:
        pre = 0
        flag = 0
        res = list()
        for i,v in enumerate(S):
            if v == '(':
                flag += 1
            else:
                flag -= 1
            
            if flag == 0:
                res.append(S[pre+1:i])
                pre = i+1

        return ''.join(res)

    def removeOuterParentheses1(self, S):
        res, opened = [], 0
        for c in S:
            if c == '(' and opened > 0: res.append(c)
            if c == ')' and opened > 1: res.append(c)
            opened += 1 if c == '(' else -1
        return "".join(res)

a = Solution()
print(a.removeOuterParentheses("()()"))

File: 1021. Remove Outermost Parentheses/1021.go
package leetcode

import(
	"strings"
)

func removeOuterParentheses(S string) string {
	pre := 0
	res := make([]string, 0)
	flag := 0
	for i,v := range S{
		if v == '('{
			flag ++
		}else{
			flag --
		}

		if flag == 0{
			res = append(res, S[pre+1: i])
			pre = i+1
		}
	}

	return strings.Join(res, "")
}

func removeOuterParentheses1(S string) string {
	var sb strings.Builder
	i, count, size := 0, 0, len(S)
	for j := 0; j < size; j++ {
		if S[j] == '(' {
			count++
		} else {
			count--
		}
		if count == 0 {
			sb.WriteString(S[i+1 : j])
			i = j + 1
		}
	}
	return sb.String()
}